GitExtUtils/GitUI/CancellableSequence.cs

```diff
@@ -0,0 +1,111 @@
+﻿using System;
+using System.Threading;
+
+namespace GitUI
+{
+    /// <summary>
+    /// Supports sequences of actions where the commencement of an action cancels any
+    /// prior action.
+    /// </summary>
+    /// <remarks>
+    /// <example>
+    /// Define an instance of this type, usually as a private readonly field:
+    /// <code>
+    /// private readonly CancellableSequence _sequence = new CancellableSequence();
+    /// </code>
+    /// Then use it to generate <see cref="CancellationToken"/> objects for use in asynchronous
+    /// operations.
+    /// <code>
+    /// var token = _sequence.Next();
+    /// // Do asynchronous operation using this token.
+    /// // Subsequent calls to Next will cancel this token.
+    /// </code>
+    /// </example>
+    /// </remarks>
+    public sealed class CancellableSequence : IDisposable
+    {
+        private CancellationTokenSource _cancellationTokenSource = new CancellationTokenSource();
+        private int _isDisposed;
+
+        /// <summary>
+        /// Issues the <see cref="CancellationToken"/> for use by the next asynchronous operation in the sequence,
+        /// and cancels any prior token.
+        /// </summary>
+        /// <remarks>
+        /// This method is thread-safe.
+        /// </remarks>
+        /// <exception cref="ObjectDisposedException">This object is disposed.</exception>
+        public CancellationToken Next()
+        {
+            if (_isDisposed != 0)
+            {
+                throw new ObjectDisposedException(nameof(CancellableSequence));
+            }
+
+            CancellationTokenSource prior;
+            CancellationTokenSource next;
+
+            do
+            {
+                prior = _cancellationTokenSource;
+
+                // Cancel any existing action in the sequence
+                prior.Cancel();
+
+                // Create a cancellation token for the next action
+                next = new CancellationTokenSource();
+            }
+            while (Interlocked.CompareExchange(ref _cancellationTokenSource, next, prior) == next);
+
+            // The above CompareExchange prevents a race condition:
+            //
+            // - Two calls to Next occur simultaneously
+            // - They both cancel the token that existed before they called
```

Comment:
This sentence is false.

Reply:
There is no guarantee about what token is the _cancellationTokenSource while the assignment, C may see B's token as well.

Reply:
There are many ways a race could pan out. The one where C cancels B is fine because it leaves only one uncancelled token. The tricky one is where they both cancel A.